### ISSUE TITLE
Stock reduced twice with one payment

### ISSUE THREAD

**Issue by thenbrent** [Automattic/woocommerce-payments#3345]
### Describe the bug

On occasion, stock is being reduced twice for a single order/payment, as discussed in this [WordPress.org thread](https://wordpress.org/support/topic/duplicate-stock-reduction-from-one-payment/).

### To Reproduce

This is sporadic so difficult to reproduce. It's potentially a race condition between webhooks and something else handling the order status - checkout logic perhaps?

At a complete guess for steps to reproduce, adding a delay between checkout processing the order may allow the webhook to consistently update it first, and it will be interesting to see if that causes it.

A code review of the stock reduction may be a better way to diagnose the cause. As one customer mentioned via ZenDesk:

> In `class-wc-payment-gateway-wcpay.php` I can see that you use wc_reduce_stock_levels when the stock gets reduced. Maybe it would be wise and solve the problem when you first check if stock has been reduced already and use wc_maybe_reduce_stock_levels instead?

### Actual behavior

Stock is reduced twice. Order note shows the order is switched twice from “Pending” to “Processing” too - example screenshot next.

### Screenshots

![2dQOJB](https://user-images.githubusercontent.com/235523/141234597-55046961-fe3e-44b2-b95e-97c485849ead.png)

### Expected behavior

Stock is reduced only once. Order status is only updated once.

### Additional context

* [WordPress.org forum thread](https://wordpress.org/support/topic/duplicate-stock-reduction-from-one-payment/)
* Ticket 4439221-zd-woothemes
* P2 for a customer affected: pcJe93-1o-p2#comment-612 - this P2 also explains why this is being prioritized high.

**Comment by debba** [comment]
Hello @thenbrent ,
I was going to report the same issue.
Today I finally discovered that this behavior occurs when user enables the option "save credit card details for faster checkouts" option.
I hope it can be helpful in solving this long-standing problem.

**Comment by thenbrent** [comment]
Thanks @debba!

**Comment by malithsen** [comment]
Just noting that I was not able to replicate this issue, after reviewing the code it looks like `wc_reduce_stock_levels` calls are not required, #3399 proposes their removal

**Comment by jessLundie** [comment]
We have another report of this in 4318150-zd-woothemes. Issue is intermittent, and is also resulting in duplicate order emails sent to customers due to the double status change.

**Comment by jessLundie** [comment]
I just took a look at that PR and it looks like that PR resolves the stock issue, but _doesn't_ address any other functionality that might be hooked into that status change.

In addition to order emails, it seems like the duplicate status change has the potential to cause issues with any 3rd party plugin that is utilizing the Pending -> Paid WooCommerce order status change - quite a few, I'd guess. 

Should I open a new issue completely to address the duplicate status change as a whole? Do we have any next steps for troubleshooting the duplicate order status change more generally?

**Comment by debba** [comment]
I can confirm sometimes the issue keeps happening, 
It happens randomly, although much less frequently than before.

**Comment by malithsen** [comment]
> Should I open a new issue completely to address the duplicate status change as a whole?

Hi @jessLundie I think we can track it under this issue. In light of new information (especially 4318150-zd-woothemes) it is likely this issue is caused at a higher-level process and the currently proposed changes in #3399 are unlikely to fix it.

> Do we have any next steps for troubleshooting the duplicate order status change more generally?

I will look further for potential causes in the code while we figure out a way to reproduce this issue. log entries from the user in 4318150-zd-woothemes will also be incredibly useful.

**Comment by malithsen** [comment]
I see that https://github.com/Automattic/woocommerce-payments/pull/3287 addresses a race condition that could lead to multiple status changes (which would then cause multiple stock-reductions/emails as reported in this issue). This change went out with v3.3.0. All instances of user reports with SSRs (4318150-zd-woothemes, 4439221-zd-woothemes, [WordPress.org thread](https://wordpress.org/support/topic/duplicate-stock-reduction-from-one-payment/)) uses versions prior to the fix.

@debba Does this issue keep happening to you on the latest version?

**Comment by malithsen** [comment]
ah thanks @billrobbins I only looked at the SSR. Looking at the status-change logs of that particular site, I can confirm that while #3287 alleviates the issue, the race condition can still happen. I can now reproduce the issue introducing a delay within `payment_complete`. Working towards a fix.

I'm also closing #3399 as it doesn't address this issue.

**Comment by nicholasjeberlin** [comment]
Looks like we are running into this as well in 33308160-hc
